# **Введение**

В этом уроке мы познакомимся с одной из самых популярных и мощных библиотек Python, предназначенной для работы с многомерными массивами и выполнения математических операций над ними. Если вы только начинаете работать с Python или хотите улучшить свои навыки, этот урок поможет вам разобраться в основных возможностях NumPy.

**Что такое NumPy?**

NumPy – это библиотека, которая широко используется для эффективной работы с числовыми данными. Она находит применение в таких областях, как анализ данных, научные вычисления, машинное обучение и других, где необходимо работать с большими объемами данных и выполнять сложные вычисления. NumPy значительно ускоряет обработку данных за счет оптимизации математических операций.

**Цель занятия -**
На этом занятии вы узнаете:

- Для каких задач подходит NumPy.
- Как создавать и управлять многомерными массивами.
- Как выполнять базовые операции с массивами, такие как сложение, умножение, транспонирование.
- Как использовать функции NumPy для упрощения работы с данными.
- Как NumPy может помочь улучшить производительность вашего кода за счет оптимизации вычислений.


**Что мы изучим?**

**Основы NumPy:**
Мы начнем с изучения того, что такое многомерные массивы и чем они отличаются от обычных списков в Python.

**Создание массивов:**
Научимся создавать массивы из списков, управлять их размерами и типами данных.

**Операции с массивами:**
Разберемся с основными операциями с массивами: сложение, умножение, транспонирование и другие математические операции.

**Индексация и срезы:**
Узнаем, как обращаться к элементам массива, извлекать подмассивы и изменять их содержимое.

**Функции NumPy:**
Познакомимся с наиболее полезными функциями библиотеки, такими как np.mean, np.sum и другими, которые значительно упрощают работу с данными.

В конце занятия мы рассмотрим практические примеры, где NumPy используется для решения реальных задач, и обсудим, как библиотека может помочь вам повысить эффективность программ. Это занятие подготовит вас к использованию NumPy в собственных проектах.



# **Зачем нужна библиотека NumPy?**

Представьте, что вы решили собрать вместе всех друзей на ужин. Каждый приносит с собой свои любимые блюда. У кого-то салат, у кого-то пироги, у кого-то напитки. Теперь вам нужно разложить все это по тарелкам и подать на стол. Если у вас всего пара гостей, это легко сделать руками. Но что если у вас 50 гостей? Тут уже потребуется специальный инструмент, чтобы облегчить задачу.

Можно сказать, что это похоже на то, как если бы все ваши друзья вдруг объединились и начали помогать вам раскладывать еду. Это, безусловно, ускорило бы процесс, ведь вместе всегда легче справиться с задачей. Так и NumPy помогает вам быстрее и эффективнее работать с числовыми данными.

NumPy — это как большой кухонный комбайн для чисел. Он позволяет быстро и эффективно выполнять сложные математические операции, сортировку, фильтрацию и многое другое. Это похоже на то, как если бы вы использовали блендер для приготовления смузи, вместо того чтобы смешивать ингредиенты вручную.

С NumPy вы можете работать с огромными объемами данных так же легко, как с несколькими числами. Это сэкономит вам массу времени и сил, ведь NumPy делает всю тяжелую работу за вас. Вместо того чтобы думать о том, как правильно организовать данные, вы можете сосредоточиться на самом важном — анализе и понимании информации.

Таким образом, если вы хотите эффективно работать с числами, NumPy — это ваш лучший друг. Он сделает вашу работу легче и быстрее, помогая вам увидеть картину целиком и принять правильные решения.



In [ ]:
# Импорт модуля рекомендуется делать вот так
import numpy as np



Для работы с большими объемами числовых данных списки Python, несмотря на свою универсальность, часто оказываются недостаточно эффективными. NumPy решает эту проблему, предоставляя оптимизированные многомерные массивы, которые позволяют выполнять вычисления значительно быстрее. Благодаря векторизации операций, NumPy выполняет математические операции над целыми массивами за одно действие, что существенно ускоряет обработку данных.


Создадим от 0 до 10млн на tuple, list, numpy

Теперь посмотрим скорость на реальных примерах


In [ ]:
%%time
my_list = list(_ for _ in range(10000000))


CPU times: user 609 ms, sys: 249 ms, total: 858 ms
Wall time: 876 ms


In [ ]:
%%time
my_list = [_ for _ in range(10000000)]


CPU times: user 378 ms, sys: 227 ms, total: 605 ms
Wall time: 607 ms


In [ ]:
%%time
my_tuple = tuple(_ for _ in range(10000000))

CPU times: user 696 ms, sys: 248 ms, total: 944 ms
Wall time: 945 ms


In [ ]:
%%time
my_list = np.arange(10000000)

CPU times: user 2.61 ms, sys: 17.9 ms, total: 20.5 ms
Wall time: 20.4 ms


На Wall time хорошо видно что NumPy быстрее минимум 10 раз , то и больше

Чтобы получить из списка либо кортежа np.array , нужно всего лишь передать его в метод np.array

In [ ]:

simple_list = [10, 20, 30, 40, 50]
simple_tuple = (10, 20, 30, 40, 50)

print(f"Одномерный массив (list)", simple_list, sep='-> ')
print(f"Одномерный массив (tuple)", simple_tuple, sep='-> ')
print("Получение из них numpy.array", np.array(simple_list), np.array(simple_tuple) )


Одномерный массив (list)-> [10, 20, 30, 40, 50]
Одномерный массив (tuple)-> (10, 20, 30, 40, 50)
Получение из них numpy.array [10 20 30 40 50] [10 20 30 40 50]


In [ ]:
numbers_int=np.array([1,4,5,2,6])
numbers_float=np.array([1,4.8,5,2.2,6])
print('тип данных первой ' ,numbers_int.dtype)
print('тип данных второй ' ,numbers_float.dtype)

тип данных первой  int64
тип данных второй  float64


# Двумерный массив в NumPy

Представь себе кинотеатр, где зрители сидят в рядах и местах, пронумерованных цифрами. Это двухмерный массив, где каждая строка соответствует ряду, а каждое место в строке — конкретному месту в этом ряду.

Предположим, что наш кинотеатр имеет 10 рядов и 15 мест в каждом ряду. В этом случае, мы можем представить это как двумерный массив размером 10x15. Строки представляют ряды, а колонки — места в одном ряду. Каждое место в театре можно представить как элемент массива, где первая цифра — номер ряда, вторая — номер места.



Двумерный массив в библиотеке NumPy представляет собой матрицу, состоящую из строк и столбцов. Создать двумерный массив можно с помощью функции np.array(), передав ей список списков:

- Первая строка: `[1, 2, 3, 4]`
- Вторая строка: `[5, 6, 7, 8]`
- Третья строка: `[9, 10, 11, 12]`

В отличие от списков, мы можем быстро изменять как все значения, так и строки или столбцы. И, конечно же, не забываем про маски, основанные на условиях, математические функции над ними. Это и многое другое мы обсудим позже, а сейчас давайте взглянем на двумерный массив, используя список в списках.


In [ ]:
two_dimensional_array = np.array([[1, 2, 3, 4],
                                  [5, 6, 7, 8],
                                  [9,10, 11,12]])


Чтобы узнать размерность массива используем метод ndim

In [ ]:
two_dimensional_array.ndim

2

Размерность по каждому измерению(shape)
В данном примере 3 строки 4 столбца

In [ ]:
two_dimensional_array.shape

(3, 4)

Количество элементов в массиве(суммарное) метод size

In [ ]:
two_dimensional_array.size

12

Доступ к конкретному элементу массива есть два способа.
- Первый



In [ ]:
two_dimensional_array[2][3]

12

- Второй

In [ ]:
two_dimensional_array[2,3]

12

# Изменение масивов



В этом примере демонстрируется работа с массивами библиотеки NumPy. Мы рассмотрим несколько операций, таких как добавление новых элементов, изменение существующих и выполнение арифметических операций и удаление элемента.





## Шаги, которые мы будем выполнять:



1) Создание исходного массива.

2) Использование функции `np.append` для добавления элементов в конец массива.

3) Проверка того, что исходный массив остался неизменным.

4) Добавление нового элемента в конец массива.

5) Вставка элемента в начало массива.

6) Изменение значения одного из элементов массива.

7) Выполнение операции над всеми элементами массива одновременно.

8) Массив после удаления третьего элемента.

9) Операция над всеми сразу без итератора.

In [ ]:

my_array = np.array([1, 2, 3, 4, 5])
print("Исходный массив:", my_array)

my_array[0] = 100
print("Массив после изменения первого элемента:", my_array)

new_array = np.append(my_array, [6, 7, 8])
print("Массив после append:", new_array)
print("Исходный массив не изменился:", my_array)

my_array = np.append(my_array, 9)
print("Добавление элемента в конец:", my_array)

my_array = np.insert(my_array, 0, 0)
print("Добавление в начало:",my_array)

my_array[5]=19
print("Изменение элемента:",my_array)

my_array = np.delete(my_array, 2)
print("Массив после удаления третьего элемента:", my_array)

my_array=my_array+1000
print("Операция над всеми сразу без итератора:",my_array)

Исходный массив: [1 2 3 4 5]
Массив после изменения первого элемента: [100   2   3   4   5]
Массив после append: [100   2   3   4   5   6   7   8]
Исходный массив не изменился: [100   2   3   4   5]
Добавление элемента в конец: [100   2   3   4   5   9]
Добавление в начало: [  0 100   2   3   4   5   9]
Изменение элемента: [  0 100   2   3   4  19   9]
Массив после удаления третьего элемента: [  0 100   3   4  19   9]
Операция над всеми сразу без итератора: [1000 1100 1003 1004 1019 1009]


###Что можно делать с помощью среза в NumPy (Slice)


Это способ извлечения подмножества элементов из последовательности (списка, кортежа, строки) или многомерного массива. Срезы позволяют выбрать часть данных на основе индексов начала, конца и шага.

Формат среза выглядит следующим образом: `sequence[start:end:step]`



 `Извлечение подмассива`: Срезы позволяют извлекать подмассивы из больших массивов, что полезно для анализа частей данных.
 `Манипуляция данными`: Можно изменять отдельные части массива, используя срезы для их выделения.
 `Обработка изображений`: В обработке изображений срезы часто используются для выделения областей интереса (ROI) или обработки отдельных каналов изображения.
 `Работа с временными рядами`: При работе с временными рядами срезы помогают выбирать определенные временные интервалы.
 `Создание новых массивов`: Срезы могут использоваться для создания новых массивов на основе существующих данных.


Зачем использовать срезы в NumPy?

`Эффективность`: Срезы позволяют работать с частями массива без создания копий, что экономит память и ускоряет вычисления.
`Гибкость`: Возможность задавать произвольные шаги и диапазоны индексов дает большую гибкость в выборе нужных данных.
`Удобство`: Синтаксис среза интуитивно понятен и прост в использовании, что облегчает работу с многомерными массивами.
Таким образом, срезы являются мощным инструментом для работы с массивами в NumPy, позволяя эффективно манипулировать данными и извлекать нужные подмножества.

In [ ]:
a = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

print(a[1, 2])  # Элемент в 2-й строке и 3-м столбце
print(a[0, :])  # Вся первая строка
print(a[:, 1])  # Весь второй столбец

6
[1 2 3]
[2 5 8]


In [ ]:
b = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]])

# Извлечение первых двух строк и первых двух столбцов
print(b[:2, :2])

[[1 2]
 [4 5]]


In [ ]:
my_array=np.arange(60).reshape(4,3,5)
my_array

array([[[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14]],

       [[15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29]],

       [[30, 31, 32, 33, 34],
        [35, 36, 37, 38, 39],
        [40, 41, 42, 43, 44]],

       [[45, 46, 47, 48, 49],
        [50, 51, 52, 53, 54],
        [55, 56, 57, 58, 59]]])

здесь мы взяли  1 и 2 из первой оси и из этих двух матриц 3x5 взяли последний второй индекс по второй оси и из этих двух одномерных массивов длинной 5 взяли до 5-го индекса каждую вторую то есть [0,2,4]

In [ ]:
my_array[1:3:1, 2, :5:2]

array([[25, 27, 29],
       [40, 42, 44]])

Теперь давайте посмотрим на несколько базовых операций с массивами:
- Сложение
- Вычитание
- Умножение
- Деление


In [ ]:
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])
print(a + b)
print(a - b)
print(a * b)
print(a / b)

[5 7 9]
[-3 -3 -3]
[ 4 10 18]
[0.25 0.4  0.5 ]




# Объединение массивов




*Когда это может пригодиться?*
- `Наука и инженерия`: Манипуляции с данными и матрицами важны при анализе экспериментальных данных, моделировании или работе с большими массивами чисел.
- `Обработка изображений` и видео: Изображения часто представляются в виде массивов (матриц), и для анализа, обработки или передачи данных эти операции необходимы для преобразования форматов.
- `Численные вычисления`: При решении систем уравнений, обработке матриц или векторах важно уметь манипулировать данными в различных формах и объединять их.

Шаги, которые мы будем выполнять:

1) Создание двух исходных массивов.

2) Объединение массивов с помощью `np.concatenate`.

3) Горизонтальное объединение массивов с помощью `np.hstack`.

4) Вертикальное объединение массивов с помощью `np.vstack`.

5) Изменение формы массива с помощью `reshape`.

6) Применение транспонирования к массиву.

7) Преобразование двумерного массива в одномерный с помощью `reshape(-1)`.

8) Преобразование двумерного массива в одномерный с помощью `flatten`.

9) Преобразование двумерного массива в одномерный с помощью `ravel`.



In [ ]:
# 1) Создание двух исходных массивов.
array1 = np.array([1, 2, 3])
array2 = np.array([4, 5, 6])
print("1) Массивы:")
print(" Массив 1:", array1)
print(" Массив 2:", array2)

# 2) Объединение массивов с помощью `np.concatenate`.
concatenated_array = np.concatenate((array1, array2))
print("\n2) Объединение массивов с помощью concatenate:", concatenated_array)

# 3) Горизонтальное объединение массивов с помощью `np.hstack`.
horizontal_stack = np.hstack((array1, array2))
print("3) Горизонтальное объединение массивов:", horizontal_stack)

# 4) Вертикальное объединение массивов с помощью `np.vstack`.
vertical_stack = np.vstack((array1, array2))
print("\n4) Вертикальное объединение массивов:")
print(vertical_stack)

# 5) Изменение формы массива с помощью `reshape`.
my_array = np.arange(12)
print("\n5) Исходный массив",my_array)

new_shape = my_array.reshape(2,6)
print("\n6) Измененная форма массива 2x6")
print(new_shape)

# 6) Применение транспонирования к массиву.
print("\n7) Применение transpose к массиву мы получаем вместо 2x6 x 6x2 размерность")
print(new_shape.transpose())

# 7) Преобразование двумерного массива в одномерный с помощью `reshape(-1)`.
print("\n8) Измененная форма массива в одномерный через reshape(-1)")
print(new_shape.reshape(-1))

# 8) Преобразование двумерного массива в одномерный с помощью `flatten`.
print("\n9) Измененная форма массива в одномерный через flatten")
print(new_shape.flatten())

# 9) Преобразование двумерного массива в одномерный с помощью `ravel`.
print("\n10) Измененная форма массива в одномерный через ravel")
print(new_shape.ravel())

1) Массивы:
 Массив 1: [1 2 3]
 Массив 2: [4 5 6]

2) Объединение массивов с помощью concatenate: [1 2 3 4 5 6]
3) Горизонтальное объединение массивов: [1 2 3 4 5 6]

4) Вертикальное объединение массивов:
[[1 2 3]
 [4 5 6]]

5) Исходный массив [ 0  1  2  3  4  5  6  7  8  9 10 11]

6) Измененная форма массива 2x6
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]]

7) Применение transpose к массиву мы получаем вместо 2x6 x 6x2 размерность
[[ 0  6]
 [ 1  7]
 [ 2  8]
 [ 3  9]
 [ 4 10]
 [ 5 11]]

8) Измененная форма массива в одномерный через reshape(-1)
[ 0  1  2  3  4  5  6  7  8  9 10 11]

9) Измененная форма массива в одномерный через flatten
[ 0  1  2  3  4  5  6  7  8  9 10 11]

10) Измененная форма массива в одномерный через ravel
[ 0  1  2  3  4  5  6  7  8  9 10 11]


# Генерация значений


Зачем нужна генерация случайных чисел?

Генерация случайных чисел является важной частью многих алгоритмов и приложений:



#### `Моделирование:`
  Случайные числа используются при создании моделей, которые имитируют реальные процессы, такие как физические явления, финансовые рынки, поведение биологических систем и многое другое.
#### `Тестирование:`
   Для тестирования программ часто требуется создание различных наборов входных данных, чтобы проверить корректность работы программы во всех возможных сценариях.
#### `Машинное обучение:`
   В машинном обучении случайные числа применяются для инициализации весов нейронной сети, перемешивания данных перед обучением модели, а также для других задач, таких как дропаут.
#### `Криптография:`
   В криптографии случайные числа необходимы для создания ключей шифрования и обеспечения безопасности данных.


#### Основные функции для генерации случайных чисел в NumPy
#### В библиотеке NumPy есть несколько функций для генерации случайных чисел:

#### `np.random.rand()`: Генерирует равномерно распределённые случайные числа от 0 до 1.
#### `np.random.randint(low, high, size=None)`: Генерирует целые случайные числа в заданном диапазоне [low, high).
#### `np.random.normal(loc=0.0, scale=1.0, size=None)`: Генерирует нормально распределённые случайные числа со средним значением loc и стандартным отклонением scale.
#### `np.random.choice(a, size=None, replace=True, p=None)`: Выбирает случайный элемент из массива a, возможно с заменой (replace) и вероятностями p.
#### `Заключение`
Случайная генерация чисел играет важную роль в науке о данных, машинном обучении, моделировании и тестировании. Библиотека NumPy предлагает удобные и эффективные методы для создания случайных массивов, что делает её незаменимой для решения широкого круга задач.


Метод arange(start, stop, step, dtype)
- `start`: начальное значение
- `stop`: конечное значение (не включается в результат)
- `step`: шаг
- `dtype`: тип данных






In [ ]:
np.arange(start=3.6, stop=10, step=0.6, dtype=float)

array([3.6, 4.2, 4.8, 5.4, 6. , 6.6, 7.2, 7.8, 8.4, 9. , 9.6])

Зачем нужен np.arange()?

Создание последовательностей чисел:
Одной из самых распространенных задач является создание последовательности чисел для дальнейших вычислений, например, для итерации по индексам, генерации временных рядов или просто для работы с числовыми данными.

Пример: создание массива индексов для работы с массивами данных.

In [ ]:
indices = np.arange(0, 10, 2)
indices

array([0, 2, 4, 6, 8])

Генерация временных рядов или шагов на интервале: Когда нужно создать значения, например, для временных рядов или данных, представленных в виде равномерно распределенных точек на некотором интервале.

Пример: создание временной шкалы для измерений с шагом 0.1.

In [ ]:
time_arr = np.arange(0, 10, 0.1)
time_arr

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. , 1.1, 1.2,
       1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2. , 2.1, 2.2, 2.3, 2.4, 2.5,
       2.6, 2.7, 2.8, 2.9, 3. , 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8,
       3.9, 4. , 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5. , 5.1,
       5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6. , 6.1, 6.2, 6.3, 6.4,
       6.5, 6.6, 6.7, 6.8, 6.9, 7. , 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7,
       7.8, 7.9, 8. , 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9. ,
       9.1, 9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9])

Удобство работы с массивами и векторизация:
В отличие от стандартного range(), который возвращает генератор, np.arange() создает сразу массив, что позволяет напрямую выполнять операции над массивами в стиле векторизации — быстрого выполнения математических операций с массивами без явных циклов.

Использование в задачах машинного обучения: В задачах машинного обучения часто нужно создать диапазоны значений для подбора гиперпараметров, генерации обучающих данных или выполнения различных операций с матрицами.

Пример: генерация диапазонов значений для настройки параметров модели.

In [ ]:
learning_rates = np.arange(0.001, 0.1, 0.01)
learning_rates


array([0.001, 0.011, 0.021, 0.031, 0.041, 0.051, 0.061, 0.071, 0.081,
       0.091])

 Если np.arange() генерирует массив с определенным шагом, то np.linspace() используется для создания массива

 с фиксированным количеством точек на заданном интервале.

Метод np.linspace(start, stop, num=50, endpoint=True, retstep=False ,dtype=None):
- `start` : Начальное значение интервала.
- `stop` : Конечное значение интервала.
- `num` : Количество элементов в результирующем массиве (по умолчанию 50).
- `endpoint` : Логическое значение, указывающее, включать ли конечное значение в результирующий массив (по умолчанию True).
- `retstep` : Логическое значение, указывающее, возвращать ли шаг между элементами (по умолчанию False).
- `dtype` : Тип данных элементов массива.

Когда это может пригодиться?
- Генерация данных для графиков — например, для построения функции на графике.
- Предобработка данных — например, создание индексов для обработки данных.
- Моделирование — создание массива значений для моделирования различных процессов (например, временные ряды).
- Численные методы — решение задач, связанных с численным интегрированием или аппроксимацией.
- Машинное обучение — создание диапазонов гиперпараметров для настройки моделей.


In [ ]:
np.linspace(5, 100, 8)

array([  5.        ,  18.57142857,  32.14285714,  45.71428571,
        59.28571429,  72.85714286,  86.42857143, 100.        ])

Методы ones и zeros

Зачем могут понадобиться массивы, состоящие из 1 и 0, и в каких ситуациях это полезно.
1. Маски и фильтрация данных
Массивы из 1 и 0 могут служить так называемыми масками для фильтрации данных. Маски — это массивы, в которых 1 (или True) означает, что элемент должен быть включен в операцию или отбор, а 0 (или False) означает, что элемент исключен.


2. Бинаризация изображений
В обработке изображений часто используются массивы, состоящие из 1 и 0, для бинаризации изображения — процесса преобразования цветного или градационного изображения в черно-белое, где 1 может означать "белый", а 0 — "черный".

In [ ]:

np.zeros((3, 2), dtype=int) ,np.ones((3, 2), dtype=np.float16)

(array([[0, 0],
        [0, 0],
        [0, 0]]),
 array([[1., 1.],
        [1., 1.],
        [1., 1.]], dtype=float16))

Можно так же получить такой же по размеру массив из нулей и единиц используя zeros_like и ones_like.

In [ ]:
my_array=np.array([[1,2,3],
                   [8,4,7]])

np.zeros_like(my_array), np.ones_like(my_array)

(array([[0, 0, 0],
        [0, 0, 0]]),
 array([[1, 1, 1],
        [1, 1, 1]]))

In [ ]:
my_array=np.array([1,2,3,
                   8,4,7])


Генерация случайными данными

Это самый быстрый метод чтобы создать массив нужного размера из тех которые мы в дальнейшем обсудим


In [ ]:

my_array=np.empty((5,4))
my_array

array([[0. , 0. , 0.3, 1. ],
       [0. , 0. , 1. , 1. ],
       [1. , 1. , 1. , 1. ],
       [1. , 0. , 0. , 1. ],
       [0.5, 0. , 0. , 1. ]])

Массив, заполненный случайными числами от 0 до 1:

In [ ]:
random_array = np.random.rand(2, 3)
random_array

array([[0.02656294, 0.71512645, 0.43101701],
       [0.06342018, 0.32403354, 0.5033114 ]])

In [ ]:
print(" Получаем матрицу из значений от 0 до 1", np.random.random([3,5]), sep="\n\n")
print("\n Получаем матрицу uniform от 4 до 10 5x5 размером",np.random.uniform(low=4, high=10, size=[5,5]), sep="\n\n")


 Получаем матрицу из значений от 0 до 1

[[0.72966597 0.13094136 0.81427928 0.13446996 0.45379566]
 [0.32017713 0.68835707 0.44405143 0.31787377 0.70335301]
 [0.94789139 0.06713471 0.48084742 0.40504447 0.9372021 ]]

 Получаем матрицу uniform от 4 до 10 5x5 размером

[[7.22081688 8.08155987 8.43405998 7.69709702 5.10060319]
 [4.82014536 8.65026761 6.6861057  4.5366395  9.4599046 ]
 [6.61801688 4.1411739  9.0005862  9.33574529 7.77061241]
 [6.7460148  5.42746709 4.89839017 9.67099801 7.0371684 ]
 [4.53278236 4.39570453 6.71038078 9.93560703 7.8705383 ]]


In [ ]:
print("\n Получаем матрицу int значений от 5 до 20 3x6 размером",np.random.randint(5,20,(3,6)),sep="\n\n")
print( "генерируем нормальное распределение loc=7, scale=2.0 ", np.random.normal(loc=7, scale=2.0, size=(3,3)))



 Получаем матрицу int значений от 5 до 20 3x6 размером

[[14 15 13 11 14  7]
 [19 12  8 10 10 17]
 [15  5  7 13 17  5]]
генерируем нормальное распределение loc=7, scale=2.0  [[ 2.63605296 10.53860561  7.00921052]
 [ 6.37162366  4.65882623  6.47070203]
 [ 7.01928619  9.51877522  6.08136366]]


In [ ]:
my_array=np.arange(20)
print('Получаем рандомные значения из указаного массива в указанном размере')
np.random.choice(my_array, size=[2,2])

Получаем рандомные значения из указаного массива в указанном размере


array([[ 5, 13],
       [ 5, 15]])

Есть массив из которого надо получить рандомные значения , это можно сделать не только передавая рандомные индексы или получая рандомные значения как это было показанно используя np.random.choice , но и через маску. В этом случае важно передавать маску такого же размера как и массив.

In [ ]:
my_array=np.random.randint(5, 10,10)
my_array

array([6, 6, 7, 5, 9, 9, 8, 6, 9, 7])

**Булевые маски**

Это масив в котором указывается True или False и беруться только те значения у которых индекс True

In [ ]:
random_mask=np.random.choice([True, False], size=len(my_array))
random_mask

array([False,  True,  True,  True, False, False, False, False, False,
       False])

In [ ]:
my_array[random_mask]

array([6, 7, 5])

Можно не только рандомно получать маски, но и использовать логические выражения. Извлечение элементов по условию (логическое индексирование)


Вы можете извлекать элементы массива на основе условия.

In [ ]:
mask=my_array%2==0
mask

array([ True,  True, False, False, False, False,  True,  True, False,
       False])

Получаем только чётные числа из массива

In [ ]:
my_array[mask]

array([6, 6, 8, 6])

In [ ]:
a = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])

# Извлекаем все элементы больше 5 или меньше 2
# Примечание этот метод не работает с ([and] ,[or]) надо ([&] , [|] )

print(a[(a > 5) | (a < 2)])

[1 6 7 8 9]


# Рассмотрим другие не менее полезные методы


- where
- агрегация данных
- смена типа данных через astype
- универсальные функции (ufuncs)
- np.dot - (скалярное произведение)
- copy (копирование массива)
- посмотрим как тип данных может изменить размерность


## Where.



Вместо того, чтобы использовать цикл для замены значений в массиве в зависимости от маски, можно воспользоваться функцией where, что делает код более удобным и эффективным. Преимущество использования where заключается в том, что она работает на уровне векторных операций и использует оптимизированные алгоритмы, благодаря чему процесс замены происходит быстрее и с меньшим использованием памяти, чем при итерации по массиву вручную в цикле.

Вот как это работает:

- Если маска содержит True, то будет выполнена замена значений в массиве согласно переданному значению.
- Если маска содержит False, то замена не происходит, и исходное значение остаётся.
Такой подход позволяет избежать явного обхода массива с использованием цикла, а вместо этого позволяет выполнить все операции в одном шаге.




В первом случае мы передаем сначала значение потом массив , а во втором с начала массив потом значение. В первом случае мы заменим чётные а во втором случае заменим нечётные

In [ ]:
arr = np.array([-1, 2, -3, 4, -5])

In [ ]:

print(np.where(arr < 0, arr, 0))

print(np.where(arr < 0, 0, arr))

[-1  0 -3  0 -5]
[0 2 0 4 0]


## Агрегация данных.

В Numpy есть множество удобных и эффективных функций, которые делают работу с данными намного проще по сравнению с обычными списками Python. Например, функции для вычисления минимальных и максимальных значений, суммы и других статистических операций, таких как min(), max() и sum(), работают точно так же, как и в обычных списках. Однако с массивами Numpy эти операции выполняются гораздо быстрее и эффективнее, потому что они оптимизированы для работы с большими объемами данных и используют низкоуровневые оптимизации.

В отличие от обычных списков, которые хранят данные в виде объектов Python и требуют большего времени на выполнение операций, массивы Numpy представляют собой фиксированные блоки памяти с однородными данными, что позволяет значительно ускорить вычисления. Например:

np.min() находит минимальное значение в массиве.
np.max() находит максимальное значение.
np.sum() суммирует все элементы массива.
Все эти функции работают аналогично тем, что есть в стандартных списках Python, но они не только быстрее, но и более удобны в использовании, особенно когда дело касается многомерных данных и работы с большими массивами.

Функции для агрегации
- 1) `max`-максимальное значение
- 2) `min`-минимальное значение
- 3) `mean`-среднее значение
- 4) `sum`-сумма всех значений
- 5) `std`-стандартное отклонение
- 6) `argmin`-минимальное значения
- 7) `argmax`-максимальное значения
- 8) `argsort`-cортировка значений по индексам по возрастанию.



In [ ]:
my_array=np.random.randint(5,20,10)
my_array

array([10, 17, 15, 19, 13, 13, 17, 16, 14, 13])

In [ ]:
print('1' , my_array.max())
print('2' , my_array.min())
print('3' , my_array.mean())
print('4' , my_array.sum())
print('5' , my_array.std())
print('6' , my_array.argmin())
print('7' , my_array.argmax())
print('8' , my_array.argsort())




1 19
2 10
3 14.7
4 147
5 2.4919871588754225
6 0
7 3
8 [0 4 5 9 8 2 7 1 6 3]


In [ ]:
my_array = np.random.randint(0, 100, size=[2,5,3])
print("Исходный массив",my_array)

Исходный массив [[[93 84 33]
  [26 85 46]
  [45 63 49]
  [85  0 97]
  [66 85 33]]

 [[23 37 19]
  [53 67 21]
  [10 43 15]
  [55 10 75]
  [93  8 81]]]



Когда мы используем параметр axis=0 в методе max() для массива, мы выполняем операцию по нахождению максимальных значений вдоль оси 0, то есть по столбцам. Это позволяет получить максимальное значение в каждом столбце, сводя данные по вертикали, при этом не затрагивая отдельные строки.
Эту операцию можно применять не только для нахождения максимума, но и для любых других агрегирующих функций, таких как сумма, среднее, стандартное отклонение и т.д. Параметр axis позволяет гибко выбирать, вдоль какой оси (строк или столбцов) будет выполнена операция. Например, axis=0 агрегирует данные по столбцам, а axis=1 — по строкам. Это дает возможность проводить разнообразные вычисления на различных уровнях структуры массива.

In [ ]:
my_array.max(axis=0)

array([[93, 84, 33],
       [53, 85, 46],
       [45, 63, 49],
       [85, 10, 97],
       [93, 85, 81]])

In [ ]:
my_array.max(axis=1)


array([[93, 85, 97],
       [93, 67, 81]])

In [ ]:
my_array.max(axis=2)

array([[93, 85, 63, 97, 85],
       [37, 67, 43, 75, 93]])

## Смена типа данных через astype


In [ ]:
my_array.astype('f')

array([[[93., 84., 33.],
        [26., 85., 46.],
        [45., 63., 49.],
        [85.,  0., 97.],
        [66., 85., 33.]],

       [[23., 37., 19.],
        [53., 67., 21.],
        [10., 43., 15.],
        [55., 10., 75.],
        [93.,  8., 81.]]], dtype=float32)

In [ ]:
my_array.astype('float')

array([[[93., 84., 33.],
        [26., 85., 46.],
        [45., 63., 49.],
        [85.,  0., 97.],
        [66., 85., 33.]],

       [[23., 37., 19.],
        [53., 67., 21.],
        [10., 43., 15.],
        [55., 10., 75.],
        [93.,  8., 81.]]])

In [ ]:
my_array.astype('str')

array([[['93', '84', '33'],
        ['26', '85', '46'],
        ['45', '63', '49'],
        ['85', '0', '97'],
        ['66', '85', '33']],

       [['23', '37', '19'],
        ['53', '67', '21'],
        ['10', '43', '15'],
        ['55', '10', '75'],
        ['93', '8', '81']]], dtype='<U21')

In [ ]:
my_array.astype(np.int64)

array([[[93, 84, 33],
        [26, 85, 46],
        [45, 63, 49],
        [85,  0, 97],
        [66, 85, 33]],

       [[23, 37, 19],
        [53, 67, 21],
        [10, 43, 15],
        [55, 10, 75],
        [93,  8, 81]]])

## Универсальные функции (ufuncs):




  NumPy предоставляет множество математических функций, которые можно применять ко всем элементам массива.
- `1)` Квадратный корень.
- `2)` Экспонента
- `3)` Произведение чисел в массиве

In [ ]:
a = np.array([1, 2, 3, 4, 5])
print(np.sqrt(a))
print(np.exp(a))
print(np.prod(a))

[1.         1.41421356 1.73205081 2.         2.23606798]
[  2.71828183   7.3890561   20.08553692  54.59815003 148.4131591 ]
120


## np.dot - (скалярное произведение)



Здесь вектор умножается на матрицу, и результатом будет новый вектор, полученный путем линейной комбинации строк матрицы с коэффициентами из вектора.

Функция np.dot очень полезна для работы с линейной алгеброй, как в случае с векторами, так и в случае с матрицами.

При умножении матрицы на вектор или другую матрицу важно учитывать их размерности. Например, если у вас есть матрица размером
3
×
6
3×6 и матрица
6
×
8
6×8, результатом умножения будет матрица размером
3
×
8
3×8, поскольку внутренние размерности (в данном случае 6) должны совпадать. Важно помнить, что операция np.dot может применяться как к скалярным произведениям векторов, так и к матричным операциям, где каждый элемент результирующей матрицы получается через линейную комбинацию строк или столбцов исходных матриц.

Это стандартная операция в линейной алгебре, которая используется, например, в решении систем линейных уравнений, в нейронных сетях и при вычислениях в машинном обучении.

векторное умножение



-здесь np.dot(a, b) вычисляет скалярное произведение, которое равно
1∗4+
2∗5+3∗6=
32

In [ ]:
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])

result = np.dot(a, b)
print(result)

32


матричное умножение

In [ ]:
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])
result = np.dot(A, B)
print(result)

[[19 22]
 [43 50]]


умножение вектора на матрицу

In [ ]:

v = np.array([1, 2])
M = np.array([[1, 2], [3, 4]])
result = np.dot(v, M)
print(result)

[ 7 10]


## copy (копирование массива)

In [ ]:
my_array=np.arange(12)
new_array=np.reshape(my_array, (4,3))
new_array

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11]])

так как обе ссылаются на одни и те же данные то если изменить в одном значение то измениться и в другом

In [ ]:
new_array[2][2]=19
new_array

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7, 19],
       [ 9, 10, 11]])

In [ ]:
my_array


array([ 0,  1,  2,  3,  4,  5,  6,  7, 19,  9, 10, 11])

чтобы избежать такой проблемы надо использовать copy

In [ ]:
my_array=np.arange(12)
my_array

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [ ]:
new_array=np.reshape(my_array, (4,3)).copy()
#Или так
# new_array=np.reshape(my_array, (4,3), copy=True)

new_array

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11]])

In [ ]:
new_array[0]=444
new_array

array([[444, 444, 444],
       [  3,   4,   5],
       [  6,   7,   8],
       [  9,  10,  11]])

In [ ]:
my_array

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [ ]:
new_array

array([[444, 444, 444],
       [  3,   4,   5],
       [  6,   7,   8],
       [  9,  10,  11]])

In [ ]:
my_array=np.random.randint(1,10,(5,4))
my_array

array([[2, 6, 3, 7],
       [2, 3, 2, 3],
       [7, 2, 3, 5],
       [2, 9, 3, 2],
       [8, 8, 4, 9]])

## посмотрим как тип данных может изменить размерность

Преобразование типов данных в программировании — это часто используемая практика, которая имеет несколько целей, несмотря на возможные потери точности, например, при преобразовании из int64 в int16. Давайте разберём, зачем и почему это может понадобиться.

`Экономии памяти`:
 Меньшие типы данных занимают меньше памяти (например, int16 вместо int64), что важно при работе с большими массивами данных.

`Ускорения работы`:
 Меньшие типы могут обрабатываться быстрее, так как занимают меньше места в кэше процессора и оперативной памяти.

`Соответствия стандартам`:
 Некоторые протоколы и форматы требуют определённых типов данных (например, 16-битные числа для изображений).

`Снижения точности`:
 Иногда потеря точности не критична, а преобразование помогает уменьшить размер данных и повысить производительность.

In [ ]:
my_array.dtype=np.int16

In [ ]:
my_array.shape

(5, 16)

In [ ]:
my_array

array([[2, 0, 0, 0, 6, 0, 0, 0, 3, 0, 0, 0, 7, 0, 0, 0],
       [2, 0, 0, 0, 3, 0, 0, 0, 2, 0, 0, 0, 3, 0, 0, 0],
       [7, 0, 0, 0, 2, 0, 0, 0, 3, 0, 0, 0, 5, 0, 0, 0],
       [2, 0, 0, 0, 9, 0, 0, 0, 3, 0, 0, 0, 2, 0, 0, 0],
       [8, 0, 0, 0, 8, 0, 0, 0, 4, 0, 0, 0, 9, 0, 0, 0]], dtype=int16)

# **Тестирование**


Цель данного теста - оценить уровень понимания базовых концепций и операций с библиотекой NumPy в Python. Вопросы охватывают следующие ключевые аспекты:

- `Создание массивов`: Проверка понимания функции np.array() и ее роли в создании массивов.
- `Операции с массивами`: Оценка знания основных арифметических операций, функций агрегации (например, np.mean(), np.max()) и манипуляций с формой массива (например, reshape(), resize()).
- `Индексация и срезы`: Проверка понимания механизмов доступа к элементам массива и их изменения.
- `Создание специальных массивов`: Оценка знания функций для создания массивов, заполненных нулями или единицами.

По результатам теста можно сделать вывод о том, насколько хорошо тестируемый:

- Понимает основные принципы работы с NumPy.
- Может создавать и манипулировать массивами.
- Знает, как выполнять различные операции с массивами.
- Может применять полученные знания для решения практических задач.

## Вопросы и ответы


**1. Как называется функция в NumPy для создания массива?**

a) `np.array()`
b) `np.create()`
c) `np.init()`
d) `np.list()`

**Правильный ответ:** a) `np.array()`


**2. Какие операции можно выполнять с массивами NumPy? Выберите несколько вариантов.**

a) Сложение массивов
b) Умножение массивов
c) Деление строк
d) Вычисление среднего значения

**Правильные ответы:** a) Сложение массивов, b) Умножение массивов, d) Вычисление среднего значения


**3. Какая из следующих операций НЕ является допустимой для массива NumPy?**

a) Сложение двух массивов одинакового размера  
b) Умножение массива на число  
c) Сложение массива с числом  
d) Конкатенация массивов разного размера  

**Правильный ответ:** d) Конкатенация массивов разного размера


**4. Какие из следующих операций выполняются с использованием стандартных функций NumPy? Выберите несколько вариантов.**

a) Нахождение медианы массива
b) Вычисление максимума и минимума массива
c) Операции с кортежами
d) Сортировка массива

**Правильные ответы:** a) Нахождение медианы массива, b) Вычисление максимума и минимума массива, d) Сортировка массива


**5. Что возвращает функция np.mean()?**

a) Сумму всех элементов массива
b) Среднее значение всех элементов массива
c) Максимальное значение массива
d) Индекс минимального элемента массива

**Правильный ответ:** b) Среднее значение всех элементов массива


**6. Какие из следующих методов можно использовать для изменения формы массива в NumPy? Выберите несколько вариантов.**

a) `reshape()`
b) `flatten()`
c) `astype()`
d) `resize()`

**Правильные ответы:** a) reshape(), b) flatten(), d) resize()


**7. Какой результат будет при выполнении следующего кода?**

import numpy as np
arr = np.array([1, 2, 3, 4])
arr[1:3] = 10
print(arr)

a) [1, 10, 10, 4]
b) [1, 2, 3, 4]
c) [1, 10, 3, 4]
d) Ошибка индексации

**Правильный ответ:** a) [1, 10, 10, 4]


**8. Что делает функция np.zeros()?**

a) Создает массив, заполненный нулями
b) Создает массив случайных чисел
c) Создает пустой массив
d) Заполняет массив значениями из диапазона

**Правильный ответ:** a) Создает массив, заполненный нулями


**9. Какие методы позволяют изменять размеры массива NumPy? Выберите несколько вариантов.**

a) `reshape()`
b) `flatten()`
c) `resize()`
d) `split()`

**Правильные ответы:** a) reshape(), c) resize()


**10. Какие функции NumPy можно использовать для работы с элементами массива? Выберите несколько вариантов.**

a) `np.sum()`
b) `np.prod()`
c) `np.max()`
d) `np.toarray()`

**Правильные ответы:** a) np.sum(), b) np.prod(), c) np.max()

## **Домашнее задание**


### Задание 1: Создание массива
- Создайте одномерный массив A из 10 элементов, содержащий числа от 1 до 10 (включительно).
- Создайте одномерный массив B из 10 элементов, содержащий числа от 10 до 20
  (включительно).
- Создайте двумерный массив размером 3x3, заполненный числами от 0 до 9.

##### Шаги для решения
   - Используйте функцию `np.arange()` для создания одномерного массива `a`.
   - Для создания двумерного массива `c` используйте функцию `np.reshape()`.

##### Ожидаемый результат
   - Одномерный массив `a`: `[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]`
   - Одномерный массив `b`: `[30, 31, 32, 33, 34, 35, 36, 37, 38, 39]`
   - Двумерный массив `c`: `[[0, 1, 2, 3, 4,5, 6, 7, 8, 9]]

### Задание 2: Индексация и срезы
- Извлеките второй элемент из одномерного массива, созданного в первом задании.
- Извлеките последние два элемента из массива (для одномерного массива).
- Извлеките второй столбец из двумерного массива, созданного во втором задании.

##### Шаги для решения
   - Применяйте индексацию для извлечения второго элемента одномерного массива `a`.
   - Используйте срезы для получения последних двух элементов и второго столбца двумерного массива `c`.

##### Ожидаемый результат
   - Второй элемент одномерного массива `a`: `1`
   - Последние два элемента одномерного массива `a`: `[8, 9]`
   - Второй столбец двумерного массива `c`: `[1, 6]`

### Задание 3: Операции с массивами
- Умножьте каждый элемент одномерного массива на 2.
- Добавьте к каждому элементу массива 5.
- Найдите сумму двух массивов A и B .

##### Шаги для решения
1. Использовать операцию умножения массива на скаляр.
2. Использовать операцию сложения массива со скаляром.
3. Использовать операцию сложения двух массивов.

##### Ожидаемый результат
- Массив `A`, умноженный на 2: `[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]`
- Массив `A`, к которому добавлено 5: `[6, 7, 8, 9, 10, 11, 12, 13, 14, 15]`
- Сумма массивов `A` и `B`: `[11, 13, 15, 17, 19, 21, 23, 25, 27, 29]`

### Задание 4: Статистические операции
- Найдите максимальное значение, минимальное значение и среднее значение в одномерном массиве A.
- Найдите сумму всех элементов двумерного массива по нулевой оси.

##### Шаги для решения
1. Использовать функции `max()`, `min()`, `mean()` для нахождения статистических характеристик.
2. Использовать метод `.sum(axis=0)` для суммирования по строке.

##### Ожидаемый результат
- Максимальное значение массива `A`: `10`
- Минимальное значение массива `A`: `1`
- Среднее значение массива `A`: `5.5`
- Сумма всех элементов двумерного массива `C` по нулевой оси: `[9, 12, 15]`

### Задание 5: Логические операции
- Используя логическое условие, получите из A все число которые делятся на 4.
- Измените все элементы массива, которые меньше больше 2 и меньше 4  на число 0.

##### Шаги для решения
1. Использовать условную фильтрацию списка.
2. Применять замену значений в списке с помощью условия.

##### Ожидаемый результат
- Числа из `A`, делящиеся на 4: `[4, 8]`
- Модифицированный массив `A`: `[1, 2, 0, 0, 5, 6, 7, 8, 9, 10]`

### Задание 6:
- Создайте одномерный массив от 0 до 1000 целых чисел и найдите его среднее значение и стандартное отклонение.

##### Шаги для решения
1. Сгенерировать массив из 1000 целых чисел.
2. Использовать функции `mean()` и `std()` для расчёта среднего значения и стандартного отклонения.

##### Ожидаемый результат
- Среднее значение большого массива: `500`
- Стандартное отклонение большого массива: около `288.67`




## Аутентичные задачи, похожие на эти задания



Данные задания напоминают реальные задачи, связанные с обработкой данных и статистическим анализом. Например:

- **Анализ временных рядов**: создание и обработка массивов данных, полученных из датчиков или финансовых рынков.
- **Обработка изображений**: работа с пиксельными данными, представленными в виде двумерных массивов.
- **Машинное обучение**: подготовка данных для моделей машинного обучения, включая нормализацию, масштабирование и фильтрацию данных.
- **Финансовый анализ**: расчет средних значений, стандартных отклонений и других статистик для оценки рисков и доходности инвестиций.



## Критерии оценивания заданий



Для оценки выполнения заданий можно использовать следующие критерии:
1. **Правильность результата**: соответствие полученного результата ожидаемому.
2. **Использование правильных методов и функций**: применение подходящих инструментов библиотеки NumPy для решения каждой задачи.
3. **Читаемость и понятность кода**: использование комментариев, осмысленных имен переменных и функций.
4. **Эффективность решений**: оптимизация времени выполнения и памяти при решении задач.

# Итоги



NumPy – это библиотека Python, специально созданная для эффективной работы с большими массивами чисел. Она значительно превосходит стандартные списки Python по скорости и функциональности.


- `Создавать многомерные массивы`: формировать структуры данных для хранения чисел.
- `Выполнять арифметические операции`: складывать, вычитать, умножать массивы.
- `Использовать индексацию и срезы` : получать доступ к отдельным элементам или подмассивам.
- `Применять различные функции`: находить максимальные и минимальные значения, среднее, сумму и т.д.
- `Изменять форму массивов`: перестраивать массивы в различные конфигурации.

Ключевые функции и концепции, которые мы рассмотрели:

- `np.array()`: создание массивов
- `np.zeros()`, np.ones(): создание массивов, заполненных нулями или единицами
- `np.arange()`: создание массива с заданным диапазоном значений
- `Индексация`: доступ к элементам по их индексам
- `Срезы`: извлечение подмассивов
- `Арифметические операции`: +, -, *, /, //, **
- `Функции`: np.max(), np.min(), np.mean(), np.sum() и многие другие

Что мы еще узнали о NumPy?:
- `Высокая производительность`: оптимизированные алгоритмы и низкоуровневые реализации обеспечивают быстрые вычисления.
- `Векторизация`: позволяет выполнять операции над целыми массивами за один шаг, что ускоряет вычисления.
- `Большой набор функций`: широкий спектр математических операций и инструментов для анализа данных.
- `Основа для других библиотек`: NumPy лежит в основе многих других популярных библиотек для работы с данными, таких как Pandas, SciPy, Matplotlib их мы в дальнейшем обсудим.



# **Дополнительная литература**


1. "Python и анализ данных", Уэс Маккинни
          - Глава 7: Работа с массивами NumPy
          - Разделы:
          - Основы NumPy
          - Создание массивов
          - Индексация и срезы
          - над массивами
          - Логические операции и булевы индексы
2. "Программирование на языке Python", Марк Лутц
          -  Часть IV: Библиотеки для научных вычислений
          -  Глава 16: NumPy: научные вычисления в Python
          -  Введение в NumPy
          -  Основные типы данных и массивы
          -  Математические операции с массивами
          -  Функции для работы с массивами
3. "Изучаем Python", Марк Лутц
          -  Часть III: Приложения и инструменты
          -  Глава 15: Использование сторонних модулей и пакетов
          -  Раздел: NumPy – библиотека для научных вычислений
4. "Искусство программирования на Python", Аль Свейгарт
          -  Глава 8: Работа с числами и массивами
          -  Разделы:
          -  Введение в NumPy
          -  Создание и индексация массивов
          -  Арифметические операции с массивами
          -  Применение функций к массивам

